.ci/pitest.groovy

```diff
@@ -0,0 +1,459 @@
+        exitCode = 0
+    }
+    else if (survivedAndUnsuppressedMutations.isEmpty()
+            && !containsOnlyStableMutations(unnecessarySuppressions)) {
+        exitCode = 0
+    }
+    else {
+        if (!survivedAndUnsuppressedMutations.isEmpty()) {
+            println "Surviving mutation(s) found:"
+            survivedAndUnsuppressedMutations.each {
+                printMutation(flag, it)
+            }
+        }
+        if (!unnecessarySuppressions.isEmpty()) {
+            println "\nUnnecessary suppressed mutation(s) found:"
+            unnecessarySuppressions.each {
+                if (!it.unstable()) {
+                    printMutation(flag, it)
+                }
+            }
+        }
+        exitCode = 1
+    }
+    return exitCode
+}
+
+/**
+ * Whether a set contains only stable mutations.
+ *
+ * @param mutations A set of mutations
+ * @return {@code true} if a set contains only stable mutations
+ */
+private static boolean containsOnlyStableMutations(Set<Mutation> mutations) {
+    final Mutation unstableMutation = mutations.find {
+        return it.unstable()
+    }
+    return unstableMutation != null
+}
+
+/**
+ * Prints the mutation according to the nature of the flag.
+ *
+ * @param flag command line argument flag to determine output format
+ * @param mutation mutation to print
+ */
+private static void printMutation(String flag, Mutation mutation) {
+    if (flag != null) {
+        println mutation.toXmlString()
+    }
+    else {
+        println mutation
+    }
+}
+
+/**
+ * Determine the difference between 2 sets. The result is {@code setOne - setTwo}.
+ *
+ * @param setOne The first set in the difference
+ * @param setTwo The second set in the difference
+ * @return {@code setOne - setTwo}
+ */
+private static Set<Mutation> setDifference(final Set<Mutation> setOne,
+                                           final Set<Mutation> setTwo) {
+    final Set<Mutation> result = new HashSet<Mutation>(setOne)
+    result.removeIf(mutation -> setTwo.contains(mutation))
+    return result
+}
+
+/**
+ * A class to represent the XML {@code mutation} node.
+ */
+record Mutation(String sourceFile, String mutatedClass,
+                String mutatedMethod, String mutator,
+                String description, String lineContent, int lineNumber,
+                boolean unstable)
+        implements Comparable<Mutation> {
+
+    private static final int LINE_NUMBER_NOT_PRESENT_VALUE = -1
+
+    @Override
+    boolean equals(other) {
+        if (this.is(other)) {
+            return true
+        }
+        if (!(other in Mutation)) {
+            return false
+        }
+
+        final Mutation mutation = (Mutation) other
+
+        if (lineNumber() != LINE_NUMBER_NOT_PRESENT_VALUE
+                && mutation.lineNumber() != LINE_NUMBER_NOT_PRESENT_VALUE) {
+            if (lineNumber() != mutation.lineNumber()) {
+                return false
+            }
+        }
```

Comment:
Why?

Reply:
Mutations in suppression files don't have a line number associated with them, but mutations in the `mutations.xml` file have a line number associated with them. The line number is stored so that it can be given to the user when there are surviving mutations. So when the line number is not present, the default value of `lineNumber` is `-1` which is equivalent to `LINE_NUMBER_NOT_PRESENT_VALUE`.

But thinking further on it, this boolean condition is not required in the `equals()` or `compareTo()` method as we don't base it on line number, I removed it but the logic of `-1` line number is still there. 

I added a Javadoc comment to warn maintainers about this.